### PR TITLE
Trigger Autorelay on NAT events

### PR DESCRIPTION
**Pull request by willscott**
Make better use of the event bus between these two.

### CODE REVIEW

File: p2p/host/relay/autorelay.go [hunk 1]
```diff
@@ -84,40 +85,43 @@ func (ar *AutoRelay) hostAddrs(addrs []ma.Multiaddr) []ma.Multiaddr {
...
-	case <-time.After(autonat.AutoNATBootDelay + BootDelay):
-	case <-ctx.Done():
-		return
-	}
+	subNatPublic, _ := ar.host.EventBus().Subscribe(new(event.EvtLocalRoutabilityPublic))
+	pubChan := subNatPublic.Out()
+	subNatPrivate, _ := ar.host.EventBus().Subscribe(new(event.EvtLocalRoutabilityPrivate))
+	priChan := subNatPrivate.Out()
+	subNatUnknown, _ := ar.host.EventBus().Subscribe(new(event.EvtLocalRoutabilityUnknown))
+	unkChan := subNatUnknown.Out()
```
Comment (Stebalien) [inline p2p/host/relay/autorelay.go]: Defer a close of the subscriptions.

**Comment by Stebalien** [comment]
Is this ready for a review?

**Comment by willscott** [comment]
yep. versions of autonat aren't there, so won't merge until there's an autonat to go.mod and CI against, but logic is ready

File: config/config.go [hunk 2]
```diff
@@ -201,6 +203,13 @@ func (cfg *Config) NewNode(ctx context.Context) (host.Host, error) {
 		}
 	}
 
+	if cfg.EnableAutoRelay || cfg.AutoNATOpts != nil {
```
Comment (Stebalien) [inline config/config.go]: What sets AutoNATOpts?

Comment (willscott) [reply config/config.go]: IPFS?
Should there be a libp2p translation of the nat options?

File: config/config.go [hunk 3]
```diff
@@ -235,11 +249,32 @@ func (cfg *Config) NewNode(ctx context.Context) (host.Host, error) {
...
+		})}
+		if hop {
+			dialerStore := pstoremem.NewPeerstore()
+			dialerHost, err := cfg.newHost(ctx, dialerStore)
+			if err != nil {
+				h.Close()
+				return nil, err
+			}
+
+			autonatOpts = append(autonatOpts, autonat.EnableService(dialerHost.Network(), false))
```
Comment (Stebalien) [inline config/config.go]: Will the host get closed when the service is closed?

Comment (Stebalien) [reply config/config.go]: How do I force AutoNAT server mode?

Comment (Stebalien) [reply config/config.go]: Can I force AutoNAT client mode?

Comment (Stebalien) [reply config/config.go]: Maybe we should have an `EnableAutoNAT(mode)` option.

File: p2p/host/relay/autorelay.go [hunk 4]
```diff
@@ -66,58 +65,55 @@ func NewAutoRelay(ctx context.Context, bhost *basic.BasicHost, discover discover
...
+				update = ar.findRelays(ctx)
 			}
-			ar.status = autonat.NATStatusPublic
-			ar.mx.Unlock()
 
-		case autonat.NATStatusPrivate:
-			update := ar.findRelays(ctx)
 			ar.mx.Lock()
-			if update || ar.status != autonat.NATStatusPrivate {
+			if ar.status != evt.Reachability && evt.Reachability != network.ReachabilityUnknown {
```
Comment (Stebalien) [inline p2p/host/relay/autorelay.go]: Nit: having two cases here is confusing. I'd just write:

```go
if update || (ar.status != evt.Reachability && evt.Reachability != network.ReachabilityUnknown) { 
```

**Pull request merged**